New command: local-gadget

alban opened the pull request:
# New command: local-gadget

This patch makes the gadgets from Inspektor Gadget available on a Linux
machine without Kubernetes:
- kubectl-gadget: Inspektor Gadget on Kubernetes
- local-gadget: Inspektor Gadget on a single Linux machine

local-gadget can be executed without parameters and it will start a
shell with auto completion. Type "help" to get the list of commands.
Commands are implemented with cobra.

local-gadget uses the runc fanotify package to know when containers
based on runc are started or stopped. Notably it works with docker
containers.

Possible use cases:

- Faster development cycle of gadgets: they no longer need to be
  deployed on Kubernetes and the compilation of a single binary is
  faster than the build of the gadget container.

- Easier to write gadget tests: we can test whether a BPF program is
  accepted by the verifier without deploying to Kubernetes.

- Some gadgets have meaningful use cases on their own on a single Linux
  machine. E.g. generate seccomp profiles with Docker.

## Compilation

```
make local-gadget
```

## Example with biolatency:

```
( echo trace biolatency t1 ; \
  echo operation t1 start ; \
  sleep 5 ; \
  echo operation t1 stop ) | sudo ./local-gadget
```

```
Output: Tracing block device I/O... Hit Ctrl-C to end.

     usecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 1        |********************                    |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 2        |****************************************|
      2048 -> 4095       : 1        |********************                    |
```

## Example with seccomp:
```
( echo trace seccomp t1 mycontainer ; \
  echo operation t1 start ; \
  echo operation t1 generate ) | sudo ./local-gadget
```

```
Output: {
  "defaultAction": "SCMP_ACT_ERRNO",
  "architectures": [
    "SCMP_ARCH_X86_64",
    "SCMP_ARCH_X86",
    "SCMP_ARCH_X32"
  ],
  "syscalls": [
    {
      "names": [],
      "action": "SCMP_ACT_ALLOW"
    }
  ]
}
```
## Testing done

Tested on Minikube with the docker driver on 5.12.6-300.fc34.x86_64.

There are a few tests for LocalGadgetManager. The tests require root.

Example of execution:
```
$ sudo unshare --net make local-gadget-tests
go test -test.v ./pkg/local-gadget-manager/... -root-test \
        -tags withebpf
=== RUN   TestBasic
--- PASS: TestBasic (0.00s)
=== RUN   TestSeccomp
State: Started
OperationError:
Output:
State: Stopped
OperationError:
Output:
--- PASS: TestSeccomp (0.19s)
=== RUN   TestCollector
time="2021-08-30T09:04:25+02:00" level=info msg="Gadget socket-collector: Using PID 919055 to retrieve network namespaces of Podname \"my-container\" in Namespace \"local\""
State: Completed
OperationError:
Output: []
State: Completed
OperationError:
Output: [
]

--- PASS: TestCollector (0.13s)
PASS
ok      github.com/kinvolk/inspektor-gadget/pkg/local-gadget-manager    0.334s
```

alban commented:
Based on other PRs that should be reviewed first:
- [ ] #234
- [ ] #242 
- [ ] #243

alban commented:
More ideas for tests enabled by local-gadget:
- [ ] Test that BPF file descriptors don't leak and are not closed inappropriately. See https://github.com/kinvolk/inspektor-gadget/pull/273
- [ ] Publish a local-gadget-test container image that could be used in Flatcar automated tests in https://github.com/flatcar-linux/mantle/blob/a85e764b5619dc35e73687e7a95edbe8eca2bc62/kola/tests/coretest/core.go#L173

blanquicet commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+	if !ok {
+		return fmt.Errorf("unknown gadget")
+	}
+	factory.Delete("gadget/" + name)
```
Testing this I realised that if any operation is performed, trace will not be created so the delete operation will "fail":

```
$ ( echo trace seccomp t1 ; echo delete t1 ) | sudo ./local-gadget
INFO[0000] Deleting gadget/t1                           
INFO[0000] Deleting gadget/t1: does not exist
```

alban replied:
I can reproduce the bug too. Maybe those "INFO" message should just be removed...

blanquicet commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+	}
+	cgroupPathV2WithMountpoint, _ := containerutils.CgroupPathV2AddMountpoint(cgroupPathV2)
+	cgroupId, _ := containerutils.GetCgroupID(cgroupPathV2WithMountpoint)
+	mntns, err := containerutils.GetMntNs(pid)
```
Nit: `mntns, _ := containerutils.GetMntNs(pid)`?

blanquicet commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+	}
+	return out
+}
+func NewManager() (*LocalGadgetManager, error) {
```
New line here?

blanquicet commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+		}
+		if len(containers) == 1 {
+			if len(containers[0].Names) > 0 {
+				containerDefinition.Podname = strings.TrimPrefix(containers[0].Names[0], "/")
```
Why you did not used the `containerDefinition.Name` and the `ContainerFilter.ContainerName` for the filtering?

https://github.com/kinvolk/inspektor-gadget/blob/e2260fae3eae0862f6b58f0d9ed55d94a7b2641a/pkg/local-gadget-manager/local-gadget-manager.go#L116-L117

Or what I missed?

alban replied:
In Kubernetes, gadgets like "dns" install a tracer on the network namespace of every pod, but not on every container if the pod has several containers.

To make the gadget "dns" work in local-gadget, I needed to put the container name in the Podname field, otherwise the dns gadget would not iterate over each container.

I could change `containerDefinition` to have a new field "NetworkNamespaceOwner" bool that would be set to true for the "pause" container in the Kubernetes world (unless they have `HostNetwork=true` in their PodSpec) and true for all Docker containers with local-kube.... in this way, the dns gadget would check that `NetworkNamespaceOwner` field, and we could stop local-gadget from misusing the `Podname` field.

blanquicet commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+	}
+	l.runcNotifier = runcNotifier
+
+	l.dockercli, _ = client.NewClientWithOpts(client.FromEnv)
```
If I understood correctly the option `[container-filter]`, maybe we could check the error here to notify that filtering will not work?

alban replied:
If the Docker API does not work, the container-filter falls back on the Container ID, so filtering can still work when the user passes the Container ID...

But I agree it is not a good semantic and should be changed.

blanquicet replied:
Understood.

I am thinking that even though we adopt `--container-id` and `--container-name` notation, if the container engine initialisation fails (In this case `client.NewClientWithOpts`), it means that `--container-name` will not work. I think we should at least notify that or what about not initialising the container engine here but directly in `AddContainer`? In the end, it is only used there. Of course, this is valid only if such initialisation is not too costly and there really exist the possibility that it fails 😄.

blanquicet reviewed (commented):
I left you some comments but so far I mostly reviewed the client side. I still need to review the stream part of the `LocalGadgetManager`, the unit tests and test better all the gadgets 😄 . 

Some general comment:
- I think `local-gadget` should be added to `.gitignored`
- What do you think of removing the following files after the compilation to keep the environment clean?
  - `pkg/gadgets/dns/tracer/bpf/dns.o`
  - `pkg/gadgets/process-collector/tracer/bpf/process-collector-with-filter.o`
  - `pkg/gadgets/process-collector/tracer/bpf/process-collector.o`
  - `pkg/gadgets/seccomp/tracer/bpf/seccomp.o`
  - `pkg/gadgets/socket-collector/tracer/itertcpv4_bpfel.o`
  - `pkg/gadgets/socket-collector/tracer/iterudpv4_bpfel.o`

- I noticed socket collector is not yet working because `GetContainersBySelector` is missing, do you plan to add that support into this PR or in a future one?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+	return nil
+}
+
+func (l *LocalGadgetManager) Show(name string) error {
+	return l.Operation(name, "")
+}
```
What about having a `func (l *LocalGadgetManager) show(tracer ...)` that is used by both, `Operation()` and `Show()`. By doing that you avoid handling `""` as a special case.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Mntns
+	}
+	return ret
+}
+
+func (l *LocalGadgetManager) LookupPIDByContainer(namespace, pod, container string) uint32 {
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		if container != c.Name {
+			continue
+		}
+		return c.Pid
+	}
+	return 0
+}
+
+func (l *LocalGadgetManager) LookupPIDByPod(namespace, pod string) map[string]uint32 {
+	ret := make(map[string]uint32)
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Pid
+	}
+	return ret
+}
```
I think all of these are duplicated in gadget tracer manager. Is it too difficult to abstract the code in a way that we don't have to duplicate them?

alban replied:
Makes sense. I'll work on this refactoring first.

alban replied:
Done in https://github.com/kinvolk/inspektor-gadget/pull/298

mauriciovasquezbernal reviewed (commented):
Some nits I found during a quick look at the code. I still need to go deeper to understand the details and make more important comments.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,409 @@
+		return fmt.Errorf("unknown trace")
+	}
+
+	factory, ok := l.traceFactories[tracer.gadget]
```
Right, it should also be an impossible situation not to have the factory at this point.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+	name := strings.TrimPrefix(tracerID, "trace_gadget_")
+	t, ok := l.tracers[name]
+	if !ok {
+		return fmt.Errorf("cannot find tracer: unknown tracer %q", tracerID)
```
Can we use this same message in all the places to indicate a tracer doesn't exist?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+func (l *LocalGadgetManager) Stream(name string, stop chan struct{}) (chan string, error) {
+	t, ok := l.tracers[name]
+	if !ok {
+		return nil, fmt.Errorf("cannot find trace %q", name)
```
See above comment.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	l.ContainerAdd(containerDefinition)
+
```
nit: extra empty line

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+				Filters: filter,
+			})
+		if err != nil {
+			fmt.Printf("Couldn't find docker container: %s", err)
```
Nothing to be done apart from logging the error? Also, should it use a logger to print as debug message?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+
+}
+
+func (l *LocalGadgetManager) RemoveContainer(notif runcfanotify.ContainerEvent) {
+	l.ContainerRemove(notif.ContainerID)
+}
```
What about using `l.ContainerRemove` directly on L298?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+	return out, nil
+}
+
+func (l *LocalGadgetManager) AddContainer(notif runcfanotify.ContainerEvent) {
```
LocalGadgetManager already has `ContainerAdd`, I think these two can be easily confused. What about calling it `AddContainerRuncNotif`?

alban replied:
Ok. And I will unexport the function too (`addContainerRuncNotif` and `RemoveContainerRuncNotif`) since they are not used outside of the package.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
```
I think `client` is to generic on this context. What avoid naming this import dockerclient to make it more specific?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+
+	runcNotifier *runcfanotify.RuncNotifier
+
+	dockercli *client.Client
```
I suppose we want to extend this to support other container engines in the future. What do think about defining that interface right now and hiding the docker details from the localmanager?

alban replied:
While trying to refactor this code, I hit a dependency problem. #312 is trying to clean up some dependency problems.

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+				Labels:        map[string]string{},
+			},
+			RunMode:    "Manual",
+			OutputMode: "Stream",
```
Why always "Stream"? What about gadget that don't support it?

blanquicet replied:
I was wondering the same and after a quick check, I think `RunMode` is not being used and `OutputMode` is used just by seccomp gadget which is currently the only gadget to support multiple output modes. Therefore, I think we should include this parameter only for seccomp and add a comment explaining all this. WDYT?

mauriciovasquezbernal commented on pkg/local-gadget-manager/local-gadget-manager.go:
```diff
@@ -0,0 +1,309 @@
+			Filter: &gadgetv1alpha1.ContainerFilter{
+				Namespace:     "local",
+				Podname:       containerFilter,
+				ContainerName: "",
```
nit: I think this can be omited.

mauriciovasquezbernal reviewed (commented):
Some more comments. 

1. Existing containers are not taken into consideration. I think you can follow the same approach we use for Kubertenes to fill the existing containers. 

2. If I try to print the sockets from all containers without passing a filter I get nothing:
```
» trace socket-collector t3
» operation t3 start
State: Completed
OperationError: 
Output: []
```

alban commented on cmd/local-gadget/main.go:
```diff
@@ -0,0 +1,317 @@
+}
+
+func execInput(input string) error {
+	rootCmd := newRootCmd()
```
I was doing that before, but then if the user sets a flag likes `--help` one time, the boolean variable for the help flag is set to true, and then the help is display for all commands, until `--help=false` is passed.

alban commented on pkg/gadgettracermanager/match/match.go:
```diff
@@ -0,0 +1,45 @@
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+)
+
+func ContainerSelectorMatches(s *pb.ContainerSelector, c *pb.ContainerDefinition) bool {
```
I moved it together with `pkg/container-collection` instead. I think it's better to avoid `containerutils` for this because containerutils contains generic low-level helpers independent of Inspektor Gadget and it is imported in an external project. Arguably, `containerutils` should be a top-level package and not part of the `gadgettracermanager` directory.

And `ContainerSelectorMatches()` is specific to Inspektor Gadget (with IG's data structures).

alban commented on cmd/local-gadget/main.go:
```diff
@@ -0,0 +1,317 @@
+		}
+
+		traceCmd = &cobra.Command{
+			Use:   "trace gadget-name trace-name [ container-filter ]",
```
Yes it is. You can start docker containers with a name like:
```
docker run --name=test-seccomp -ti --rm ubuntu
```

But I think we should find a better syntax for filtering on different criteria.

Maybe `trace gadget-name trace-name --container-id 012345abcd --name test-seccomp`?

blanquicet replied:
Yes, I think `--container-id` and `--container-name` is more descriptive. @mauriciovasquezbernal, WDYT?

alban commented:
The GitHub Action fails because BTF is not available:
```
OperationError: Failed to start seccomp tracer: failed to create BPF collection:
program tracepoint__raw_syscalls__sys_enter: CO-RE relocations:
no BTF for kernel version 5.8.0-1041-azure: not supported
```
See https://github.com/kinvolk/inspektor-gadget/issues/264

Should the test be disabled when BTF is not available?

alban commented:
Branch updated with additional test helpers to detect fd leaks and go routine leaks. Those happens e.g. due to issue described in #330.

alban commented:
I rebased, renamed "trace" to "create", and removed the unsupported gadgets from local-gadget for now (they can be added back in future PR).

@eiffel-fl 
* I filed a separate PR for the crash in the seccomp gadget that you reported, see #368.
* For your 2 patches, could you file PRs for them? Thanks.

mauriciovasquezbernal commented:
I tested and it works fine in Ubuntu. I'm wondering how we're going to handle the dependency of external BCC tools like biolatency. It seems to be a difficult problem to handle, some possible problems are:
- The tools could not be installed at all. 
- The path of the tool could be different. (Already hit by @eiffel-fl above)
- The version of the tool could not be the right one.
- If we add support for a tool that we have modified in our fork then the behavior could also be different. 

Perhaps local gadget should be shipped with an installer that puts all the needed tools in a path that can be used by it? 

Besides that I think we can merge it. There are still some things to improve but we can do it on the way.

alban commented:
Branch updated:
* Rebased
* Removed changes to binary files `pkg/gadgets/socket-collector/tracer/iter*_bpfel.o`
* Remove biolatency from local-gadget

Let's see if the CI is green now

alban commented:
CI is green after updating the BTFHub URL (see also https://github.com/kinvolk/inspektor-gadget/pull/382)

mauriciovasquezbernal reviewed (approved):
I still have some questions regarding how we want distribute this binary with the files it requires. Like BCC gadgets and BTF files (https://github.com/kinvolk/inspektor-gadget/pull/387). But I think we can merge now and iterate later on.
